feat: add OAuth client developer settings page with approval workflow

PeerRich opened the pull request:
## What does this PR do?

Adds a new OAuth client management system with developer settings page and admin approval workflow.

- Fixes #26469

### Key Changes

**Developer Settings Page** (`/settings/developer/oauth`)
- New page for developers to create and manage their OAuth clients
- Clients start in PENDING status and require admin approval
- Developers can edit their clients (changes to sensitive fields trigger re-approval)

**Admin Approval Workflow** (`/settings/admin/oauth`)
- Admins can view pending, approved, and rejected OAuth clients
- Approve/reject clients with rejection reason support
- Email notifications sent to client owners on approval/rejection

**Authorization Flow Updates**
- OAuth authorization now validates client status (only APPROVED clients can authorize)
- Added `redirect_uri` validation during authorization
- Improved error handling with proper OAuth error redirects

**Database Changes**
- Added `status` field to `OAuthClient` model (PENDING, APPROVED, REJECTED)
- Added `rejectionReason`, `purpose`, `websiteUrl` fields
- Migration included

## Visual Demo (For contributors especially)

N/A - Backend/API changes with minimal UI additions

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). N/A - internal feature
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

1. **Developer flow**: Log in as a regular user, go to Settings > Developer > OAuth, create a new OAuth client. Verify it shows as "Pending".

2. **Admin flow**: Log in as admin, go to Settings > Admin > OAuth. Verify pending clients appear. Test approve/reject actions.

3. **Authorization flow**: Try to authorize with a pending client - should fail. Approve the client, then authorization should work.

4. **E2E tests**: Run `PLAYWRIGHT_HEADLESS=1 yarn e2e oauth-client-admin.e2e.ts`

## Human Review Checklist

- [ ] Verify OAuth authorization only allows APPROVED clients (security critical)
- [ ] Review status transition logic in `updateClient.handler.ts` - owner edits to sensitive fields should trigger re-approval
- [ ] Verify folder rename from `oAuth` to `oauth` is correct (fixes 404 on case-sensitive filesystems)
- [ ] E2E test uses retry logic for admin page navigation - verify this approach is acceptable for handling session timing
- [ ] Email notification triggers are correct (only on admin approve/reject actions)

## Updates since last revision

- **Fixed folder case sensitivity issue**: Renamed `admin/oAuth/` to `admin/oauth/` to fix 404 errors on case-sensitive filesystems (Linux). The code referenced `/settings/admin/oauth` (lowercase) but the folder was named `oAuth` (capital A).
- **Improved E2E test navigation**: Added `navigateToAdminOAuthPage()` helper with retry logic that waits for the admin page content to load, handling session propagation timing issues in CI.

---

Link to Devin run: https://app.devin.ai/sessions/517d12a407d34bd6b0c7855993df9dc0
Requested by: lauris@cal.com

devin-ai-integration[bot] committed:
feat: add OAuth client developer settings page with approval workflow

- Add new developer OAuth page at /settings/developer/oAuth for users to submit OAuth client requests
- Transform admin OAuth page into management dashboard for reviewing/approving submissions
- Add OAuthClientApprovalStatus enum (PENDING, APPROVED, REJECTED) to track submission status
- Add userId and createdAt fields to OAuthClient model for tracking submissions
- Create email notifications for admin (new submission) and user (approval)
- Add sidebar navigation link in developer section below API keys
- Add comprehensive translations for new UI strings
- Create OAuthClientRepository for data access following repository pattern

Co-Authored-By: peer@cal.com <peer@cal.com>

devin-ai-integration[bot] committed:
fix: re-export generateSecret for backward compatibility

Co-Authored-By: peer@cal.com <peer@cal.com>

devin-ai-integration[bot] committed:
feat: make logo mandatory and list items clickable for OAuth clients

Co-Authored-By: peer@cal.com <peer@cal.com>

devin-ai-integration[bot] committed:
fix: add missing translation keys and remove client secret from details dialog

Co-Authored-By: peer@cal.com <peer@cal.com>

devin-ai-integration[bot] committed:
fix: address cubic AI reviewer comments

- Remove duplicate 'there' JSON key in common.json
- Add select clause to findByUserId to avoid exposing clientSecret
- Add @@index([userId]) to OAuthClient model for query performance
- Update migration to include the index

Co-Authored-By: peer@cal.com <peer@cal.com>

eunjae-lee reviewed (commented):
seems good overall. i haven't tested it yet. left some nits.

@CarinaWolli can i ask you to test this PR? 🙏🏻

devin-ai-integration[bot] committed:
fix: address PR review comments - fix indentation and use useCopy hook

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

devin-ai-integration[bot] committed:
fix: change react-dom/server import to fix Turbopack compatibility

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

devin-ai-integration[bot] committed:
Revert "fix: change react-dom/server import to fix Turbopack compatibility"

This reverts commit c3e0b709c2d88fd221143cb4ce9cd25bb8c94277.

devin-ai-integration[bot] committed:
fix: use email service pattern for OAuth client notifications

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

devin-ai-integration[bot] committed:
fix: add try-catch around email sending to handle Turbopack react-dom/server issue

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

devin-ai-integration[bot] committed:
Revert "fix: add try-catch around email sending to handle Turbopack react-dom/server issue"

This reverts commit fc9d47cd773505ebc5ee2696718aad4a8a98be77.

devin-ai-integration[bot] committed:
fix: improve OAuth client UI with skeleton loaders and smaller dialog styling

- Replace 'Loading...' text with proper skeleton loaders in both developer and admin OAuth client views
- Make client_id and copy button smaller in dialogs using size='sm' and text-sm styling
- Add 'client_id' translation key to common.json for proper i18n

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

devin-ai-integration[bot] committed:
fix: improve skeleton loader to match actual OAuth client list structure

- Remove divide-y from container and use conditional border-b on rows
- Match the exact structure from oauth-clients-view.tsx L126-160
- Use proper spacing for text elements (mt-1 instead of space-y-2)

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

eunjae-lee committed:
fix skeleton

eunjae-lee committed:
rename the selected oauth client dialog

devin-ai-integration[bot] committed:
fix: address PR feedback - admin auth, dropdown styling, sidebar label

- Add defense-in-depth admin authorization check in updateClientStatus handler
- Fix broken dropdown menu by using DropdownItem with StartIcon prop
- Fix sidebar menu label from 'oAuth' to 'oauth_clients' to match developer view

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

eunjae-lee committed:
update common.json

CarinaWolli reviewed (changes requested):
We need to show the client secret to the user once the oAuth client is approved

devin-ai-integration[bot] committed:
feat: show client secret in approval email for confidential OAuth clients

- Add regenerateSecret method to OAuthClientRepository
- Regenerate secret when admin approves a PENDING confidential client
- Include client secret in approval notification email
- Add one-time warning message about storing the secret securely
- Only regenerate on first approval (not re-approvals)

Co-Authored-By: eunjae@cal.com <hey@eunjae.dev>

Ryukemeister reviewed (dismissed):
overall working as expected, couple of issues that I noticed so far:
1. as @CarinaWolli mentioned we don't display client secret even once after approval from admin. also I dont really understand the concept of just showing the client secret once, why not let the user access the client secret whenever they want to?
2. when creating a new oauth client, the form looks a bit weird. can we please fix this styling issue

<img width="3024" height="1964" alt="Image" src="https://github.com/user-attachments/assets/ba896999-7512-4c0e-a958-df46606204eb" />
3. also need to resolve the merge conflicts

supalarry reviewed (commented):
Should we also request developers to provide a Homepage URL? We can keep it optional (allowing localhost for dev), but having this allows us to display a link on the OAuth consent screen to build user trust. It also gives us better visibility into who is building on our platform. @CarinaWolli thoughts?

PS then a reminder that we need to update oauth docs https://cal.com/help/apps-and-integrations/oauth about this once live

devin-ai-integration[bot] committed:
feat: add Website URL field, fix logo styling, show client secret after approval

- Add Website URL field to OAuth client forms (admin and developer views)
- Fix Upload Logo section styling by wrapping in Label div with proper gap
- Display client secret in dialog after admin approves a confidential OAuth client
- Add websiteUrl field to Prisma schema with migration
- Update tRPC handlers and repository to support websiteUrl
- Add translation keys for new UI elements

Co-Authored-By: peer@cal.com <peer@cal.com>

devin-ai-integration[bot] committed:
fix: move clientSecret variable declaration outside if block for proper scoping

Co-Authored-By: peer@cal.com <peer@cal.com>

supalarry commented on packages/trpc/server/routers/viewer/oAuth/updateClientStatus.handler.ts:
```diff
@@ -0,0 +1,67 @@
+  if (status === "APPROVED" && clientWithUser?.user) {
+    const t = await getTranslation("en", "common");
+
+    // Only regenerate secret for confidential clients that are being approved for the first time
```
There is no need to regenerate secret - the secret is shown once to the developer when oauth client is created.

supalarry commented on packages/trpc/server/routers/viewer/oAuth/addClient.handler.ts:
```diff
@@ -1,37 +1,26 @@
 import type { TAddClientInputSchema } from "./addClient.schema";
 
+// Re-export generateSecret for backward compatibility
+export { generateSecret };
```
We need it because if you search in vs code "addclient.handler" then there are multiple files that import generateSecret from here. Instead of re-exporting we need to update those import paths to import generate secret from `"@calcom/lib/server/repository/oAuthClient";`

supalarry replied:
I removed this re-export - it was needed before indeed but not generateSecret has been moved to utils folder and all code imports from there

supalarry reviewed (changes requested):
Review (I will implement it) :

supalarry committed:
Merge branch 'main' into devin/oauth-developer-settings-1764693294

supalarry committed:
refactor: dont expose client secret in emails

supalarry commented on packages/emails/src/templates/OAuthClientApprovedNotificationEmail.tsx:
```diff
@@ -0,0 +1,112 @@
+          </tr>
+          {clientSecret && (
+            <tr style={{ lineHeight: "24px" }}>
+              <td style={{ padding: "12px", fontWeight: 600 }}>{language("client_secret")}</td>
```
I stopped sending client secret in email in commit [refactor: dont expose client secret in emails](https://github.com/calcom/cal.com/pull/25556/commits/80dc2ede5da523fd93874e8f5dd90e3c70a74353)

supalarry commented on apps/web/app/(use-page-wrapper)/settings/(settings-layout)/SettingsLayoutAppDirClient.tsx:
```diff
@@ -149,7 +150,7 @@ const getTabs = (orgBranding: OrganizationBranding | null) => {
         { name: "organizations", href: "/settings/admin/organizations", trackingMetadata: { section: "admin", page: "organizations" } },
         { name: "lockedSMS", href: "/settings/admin/lockedSMS", trackingMetadata: { section: "admin", page: "locked_sms" } },
-        { name: "oAuth", href: "/settings/admin/oAuth", trackingMetadata: { section: "admin", page: "oauth" } },
+        { name: "oauth_clients", href: "/settings/admin/oAuth", trackingMetadata: { section: "admin", page: "oauth" } },
```
i left it as before this pr

supalarry committed:
Merge branch 'main' into devin/oauth-developer-settings-1764693294

supalarry committed:
refactor: dont regenerate secret upon status change

supalarry committed:
refactor: reuse existing hash function

supalarry committed:
refactor: rename admin/oAuth to admin/oauth page

supalarry commented on apps/web/app/(use-page-wrapper)/settings/(admin-layout)/admin/oAuth/page.tsx:
```diff
@@ -2,12 +2,12 @@ import { _generateMetadata, getTranslate } from "app/_utils";
 import SettingsHeader from "@calcom/features/settings/appDir/SettingsHeader";
 
-import LegacyPage from "~/settings/admin/oauth-view";
+import OAuthClientsAdminView from "~/settings/admin/oauth-clients-admin-view";
```
Renamed in [refactor: rename admin/oAuth to admin/oauth page](https://github.com/calcom/cal.com/pull/25556/commits/c384a37d198ec75a5cadb7f873701da01eaeaaa3)

supalarry committed:
Merge branch 'main' into devin/oauth-developer-settings-1764693294

supalarry commented on apps/web/public/static/locales/en/common.json:
```diff
@@ -998,11 +998,60 @@
-  "create_oauth_client": "Create OAuth Client",
-  "create_oauth_client_description": "Create a new OAuth client for third-party integrations",
-  "oauth_client_deletion_message": "OAuth client deleted successfully",
+    "oauth_clients": "OAuth Clients",
```
its indented now

supalarry commented on apps/web/modules/settings/admin/oauth-clients-admin-view.tsx:
```diff
@@ -0,0 +1,350 @@
+                        )}
+                        <DropdownMenuItem
+                          onClick={() => {
+                            navigator.clipboard.writeText(client.clientId);
```
implemented

supalarry commented on packages/trpc/server/routers/viewer/oAuth/listUserClients.handler.ts:
```diff
@@ -0,0 +1,17 @@
+
+  const oAuthClientRepository = await OAuthClientRepository.withGlobalPrisma();
+
+  return oAuthClientRepository.findByUserId(userId);
```
resolved

supalarry committed:
refactor: deduplicate oauth repositories

supalarry committed:
refactor: remove withGlobalPrisma from oauth repository

supalarry committed:
Merge branch 'main' into devin/oauth-developer-settings-1764693294

supalarry committed:
refactor: developer oauth page

supalarry committed:
refactor: oauth status by default accepted

supalarry committed:
refactor: request oauth status when creating

supalarry committed:
refactor ux

devin-ai-integration[bot] committed:
fix: address Cubic AI code review feedback

- Add purpose field to plain text email body for accessibility
- Convert NewOAuthClientButton to inline JSX to avoid React anti-pattern
- Trigger re-approval when redirectUri changes for security
- Add e.preventDefault() for Space key to prevent page scroll
- Change default approvalStatus to PENDING for defense-in-depth
- Use oauth_clients translation key for consistency
- Add meaningful alt text to Avatar for accessibility
- Remove onClick from DialogClose to prevent double-run close effects
- Return NOT_FOUND for non-owner delete to prevent resource enumeration

Co-Authored-By: unknown <>

supalarry committed:
Merge branch 'main' into devin/oauth-developer-settings-1764693294

supalarry committed:
common.json file

supalarry committed:
refactor: delete all prisma migrations

supalarry commented on packages/trpc/server/routers/viewer/oAuth/submitClient.schema.ts:
```diff
@@ -0,0 +1,30 @@
+  redirectUri: z.string().url("Must be a valid URL"),
+  logo: z.preprocess(
+    (value) => (typeof value === "string" && value.trim() === "" ? undefined : value),
+    z.string().optional()
```
The PR description is outdated - we only require name, redirect uri and purpose. rest can be optional.

supalarry committed:
refactor: have just 1 prisma migration

supalarry commented on apps/web/modules/settings/oauth/OAuthClientFormFields.tsx:
```diff
@@ -0,0 +1,165 @@
+        <div className="space-y-3">
+          <div className="flex items-center gap-4">
+            <Avatar
+              alt=""
```
This was stupid - there is no client_logo string in en/common.json. But whatever i just switched to "logo".

supalarry commented on apps/web/app/(use-page-wrapper)/settings/(settings-layout)/SettingsLayoutAppDirClient.tsx:
```diff
@@ -127,6 +127,11 @@ const getTabs = (orgBranding: OrganizationBranding | null) => {
           trackingMetadata: { section: "developer", page: "api_keys" },
         },
+        {
+          name: "oAuth",
```
I am reverthing this - i want just "OAuth" for both devs and admin.

supalarry committed:
revert: some devin changes

supalarry committed:
fix: typecheck

supalarry committed:
test: owner OAuth crud

supalarry committed:
test: admin OAuth approval / rejection

supalarry committed:
Merge branch 'main' into devin/oauth-developer-settings-1764693294

devin-ai-integration[bot] committed:
fix: address Cubic AI review feedback (confidence 9/10 issues)

- schema.prisma: Remove @default("") from purpose field to make it required
- schema.prisma: Use UTC-aware timezone expression for createdAt default
- OAuthClientFormFields.tsx: Localize redirect URI placeholder using t()
- common.json: Add redirect_uri_placeholder translation key

Co-Authored-By: unknown <>

Commit:
cubic changes

supalarry committed:
refactor: dont log sensitive info and rethrow error

supalarry commented on packages/prisma/schema.prisma:
```diff
@@ -1769,15 +1770,30 @@ enum OAuthClientType {
+  clientSecret    String?
+  clientType      OAuthClientType           @default(CONFIDENTIAL)
+  name            String
+  purpose         String                    @default("")
```
I will handle this myself - consider this done and dont comment me about this anymore

supalarry committed:
cubic feedback

supalarry committed:
refactor: make oauth client purpose optional

supalarry committed:
refactor: admin/oauth not allowed if not logged in

supalarry committed:
refactor: admin view skeleton

supalarry committed:
refactor: rename state

supalarry committed:
refactor: get rid of redundant mapping

supalarry committed:
refactor: remove redundant handler

supalarry committed:
refactor: remove redundant handler

supalarry committed:
refactor: re-usable new oauth client button

supalarry committed:
refactor: dialogs

supalarry committed:
refactor: modals

supalarry committed:
refactor: handler names, dialog, skeleton

supalarry committed:
fix: purpose being null

supalarry committed:
refactor: rename handler and delete old oauth admin page

supalarry committed:
Merge branch 'main' into devin/oauth-developer-settings-1764693294

supalarry committed:
fix: purpose in submission

supalarry committed:
refactor: handler names

supalarry committed:
refactor: rename

supalarry committed:
refactor: update handler

supalarry committed:
refactor: rename approvalStatus -> status

supalarry committed:
refactor: simplify modal

supalarry committed:
refactor: name

supalarry committed:
dont require repproval if redirectUri changes

supalarry committed:
Merge branch 'main' into devin/oauth-developer-settings-1764693294

supalarry commented on packages/trpc/server/routers/viewer/oAuth/updateClient.handler.test.ts:
```diff
@@ -0,0 +1,488 @@
+        data: {
+          redirectUri: updatedRedirectUri,
+        },
+      })
```
are you sure? another ai replied:
```
Is that PR comment true?
No — not as written.

Even though the assertion uses expect.objectContaining(...) at the top level, your nested data assertion is a plain object:

ts
data: {
  redirectUri: updatedRedirectUri,
},
In Vitest/Jest, objectContaining only does “subset matching” for the object you pass it. For nested properties, it uses normal deep equality unless you also use matchers there.

So if the handler called Prisma like:

ts
prisma.oAuthClient.update({
  where: { clientId: CLIENT_ID },
  data: { redirectUri: updatedRedirectUri, status: "PENDING" },
})
then data would NOT deep-equal { redirectUri: updatedRedirectUri }, and the test would fail.

When would the comment be true?
It would be true if you had written:

ts
data: expect.objectContaining({ redirectUri: updatedRedirectUri })
because that would allow extra keys like status.

Optional improvement (if you want it extra explicit)
You could also assert directly against the call arg:

Assert exact data object (strongest)
Or add a negative assertion for status: "PENDING" (more intention-revealing)
But strictly speaking, your current test already detects the regression described.

Status
Answered: the claim “this test never fails if status is still set” is false given the current assertion shape.
```

supalarry committed:
Merge branch 'main' into devin/oauth-developer-settings-1764693294

supalarry committed:
fix: remove integration sync index creation

supalarry committed:
refactor: require re-approval if redirectUri updated

supalarry committed:
fix: flaky e2e test

supalarry committed:
fix: flaky e2e test

supalarry committed:
fix: flaky e2e test

CarinaWolli reviewed (changes requested):
The `authorize-view` has this code:

```
  const { data: client, isPending: isPendingGetClient } = trpc.viewer.oAuth.getClient.useQuery(
    {
      clientId: client_id as string,
    },
    {
      enabled: status !== "loading",
    }
  );
```

In the `getClient` handler, wouldn't we need to check if the client is approved? I don't think we do that yet

Same in `generateAuthorizationCode`

supalarry committed:
Merge branch 'main' into devin/oauth-developer-settings-1764693294

supalarry committed:
fix: remove duplicate common.json keys

supalarry commented on apps/web/public/static/locales/en/common.json:
```diff
@@ -1038,12 +1036,83 @@
+  "oauth_client_pending_approval": "Your OAuth client is pending approval. You will receive an email notification once it has been reviewed.",
+  "oauth_client_rejected": "Your OAuth client submission was rejected. Please contact support for more information.",
+  "oauth_client_approved_note": "Your OAuth client is approved and ready to use. The client secret was shown only once when the client was approved.",
+  "logo": "Logo",
```
Removed in [fix: remove duplicate common.json keys](https://github.com/calcom/cal.com/pull/25556/commits/ce0fb75f964e8b137919072c2b9a461aef42e51f)

supalarry committed:
refactor: replace team@cal.com with SUPPORT_MAIL_ADDRESS

supalarry committed:
refactor: generate client secret on handler level

supalarry committed:
fix: authorization code only available to approved clients

supalarry committed:
Merge branch 'main' into devin/oauth-developer-settings-1764693294

supalarry committed:
refactor: cubic review dont display exclamation

supalarry committed:
refactor: cubic review website_url in common json

supalarry committed:
fix: dont default in ui to approved status

CarinaWolli reviewed (commented):
Have some more feedback, let me know what you think:

- [ ] I think we are still missing the status check when loading the consent screen (`authorize-view.tsx`).  `trpc.viewer.oAuth.getClient.useQuery` should probably be `getApprovedClient` 

- [ ] Changing the logo shouldn't put it back to "pending", I think only changing the redirect uri should. And when making that change we should have a warning that the client won't work anymore after submitting this change

- [ ] I think also the refreshToken endpoint should stop working when the client is rejected

- [x] still can't open the admin oauth tab (https://github.com/calcom/cal.com/pull/25556/#issuecomment-3759899998), changing the folder name from `oAuth` to `oauth` works

supalarry committed:
refactor: optiona logo in schema create handler

supalarry committed:
fix: tests

supalarry committed:
fix: tests

supalarry committed:
fix: /authorize redirect if client not approved or show error

supalarry committed:
Merge branch 'main' into devin/oauth-developer-settings-1764693294

supalarry committed:
test: authorize page with invalid client id

supalarry committed:
refactor: dont allow refreshing tokens unless approved client

supalarry committed:
fix: flaky e2e test

supalarry committed:
fix: flaky e2e test

supalarry committed:
fix: flaky e2e test

supalarry committed:
fix: flaky e2e test

supalarry committed:
fix: flaky e2e test

supalarry committed:
fix: flaky e2e test

CarinaWolli commented on apps/web/modules/auth/oauth2/authorize-view.tsx:
```diff
@@ -32,15 +33,38 @@ export default function Authorize() {
+  // client is found and redirect URI is valid.
+  const canRedirectToOAuthError = getClientError?.data?.code === "UNAUTHORIZED";
+
+  useEffect(() => {
```
Not exactly exactly sure where this happens, but for pending clients we always redirect to the not yet approved redirect uri. I don't think that should happen, we should stay on cal.com 

That's where it redirects me to: 
`https://example.com/callback?error=unauthorized_client&error_description=OAuth+client+is+not+approved&state=CoX8eBF8n82HxzdUFoSA4QGhYJnBERh1MOB9m_JhD4Q`

CarinaWolli reviewed (commented):
- Can we link the docs in the Approval email? 

- You should be able to make changes to rejected OAuth clients and then click "Resubmit"

supalarry committed:
chore: warn that pending client is not usable

supalarry committed:
fix: approve and reject buttons

supalarry committed:
fix: /authorize show error if client not approved

supalarry committed:
refactor: info message about editing oauth client and status

supalarry committed:
Merge branch 'main' into devin/oauth-developer-settings-1764693294

Commit:
change info alert to warning

CarinaWolli commented:
Some e2e tests are failing, but everything else looks good from my side 🙏

supalarry committed:
try to fix ci test

supalarry committed:
debug: failing e2e test

devin-ai-integration[bot] committed:
fix: improve session propagation in oauth-client-admin E2E test

- Add navigateToAdminOAuthPage helper that waits for listClients API call
- If the API call doesn't arrive (session issue), reload page to force session refresh
- This fixes the CI flakiness where admin page wasn't loading due to session not having ADMIN role

Co-Authored-By: lauris@cal.com <lauris@cal.com>

devin-ai-integration[bot] committed:
fix: register waitForResponse before navigating in E2E test

- Register the listClients waitForResponse promise BEFORE page.goto()
- This ensures the response isn't missed during page load
- Also register the promise before reload in the catch block

Co-Authored-By: lauris@cal.com <lauris@cal.com>

devin-ai-integration[bot] committed:
fix: rename oAuth folder to oauth for case-sensitive filesystems

The admin OAuth page route was at /settings/admin/oAuth (capital A) but the
code references /settings/admin/oauth (lowercase). This caused 404 errors
on case-sensitive filesystems (Linux).

Also improved the E2E test navigation helper to retry with delays if the
admin page doesn't load immediately, handling session propagation timing.

Co-Authored-By: lauris@cal.com <lauris@cal.com>

ThyMinimalDev committed:
Merge branch 'main' into devin/oauth-developer-settings-1764693294

supalarry committed:
test style

ThyMinimalDev reviewed (approved):
LGTM
